### PR TITLE
7448_reference to resources

### PR DESCRIPTION
**Pull request by DanielWilczek**
robot side of this task
https://github.com/G1ANT-Private/G1ANT.Robot/pull/58

### CODE REVIEW

**Comment by DanielWilczek** [comment]
> putty, telnet and vnc should be embedded in the addon and copied to Documents\G1ANT.Robot\Add-on folder as it is implemented eg for selenium drivers or G1ANT.CSharpCompiler.dll

done

File: G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs [hunk 1]
```diff
@@ -27,5 +26,37 @@ namespace G1ANT.Addon.Net
...
+            foreach (var exe in exeList)
+            {
+                try
+                {
+                    using (FileStream stream = File.Create(Path.Combine(unpackfolder, exe.Key)))
+                    {
+                        stream.Write(exe.Value, 0, exe.Value.Length);
+                    }
+                }
+                catch { }
```
Comment (przegwar) [inline G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: Why this catch is empty?

Comment (DanielWilczek) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: No idea, its copy pasted from other addon.
I can only guess that we don't like to show out inconvenience :)

File: G1ANT.Addon.Net/G1ANT.Addon.Net/Commands/AS400OpenCommand.cs [hunk 2]
```diff
@@ -23,10 +23,10 @@ public class Arguments : CommandArguments
             public TextStructure Host { get; set; }
 
         }
-        public string pathToTelnet = System.IO.Path.Combine(Environment.CurrentDirectory,
-             @"..\..\..\G1ANT.Robot\Resources\telnet.exe");
-        public string pathToPutty = System.IO.Path.Combine(Environment.CurrentDirectory,
-             @"..\..\..\G1ANT.Robot\Resources\putty.exe");
+        public string pathToTelnet = System.IO.Path.Combine(AbstractSettingsContainer.Instance.UserDocsAddonFolder.FullName,
+             @"telnet.exe");
+        public string pathToPutty = System.IO.Path.Combine(AbstractSettingsContainer.Instance.UserDocsAddonFolder.FullName,
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/G1ANT.Addon.Net/Commands/AS400OpenCommand.cs]: Remove namespaces from class usages

File: G1ANT.Addon.IExplorer/G1ANT.Addon.IExplorer.Tests/G1ANT.Addon.IExplorer.Tests.csproj [hunk 3]
```diff
@@ -80,7 +80,7 @@
     <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
   </ItemGroup>
   <ItemGroup>
-    <ProjectReference Include="..\G1ANT.Addon.IE\G1ANT.Addon.IExplorer.csproj">
+    <ProjectReference Include="..\G1ANT.Addon.IExplorer\G1ANT.Addon.IExplorer.csproj">
       <Project>{cea69d5a-91c9-43d9-8928-f41d69e85baf}</Project>
       <Name>G1ANT.Addon.IExplorer</Name>
     </ProjectReference>
```
Comment (lfronczyk) [inline G1ANT.Addon.IExplorer/G1ANT.Addon.IExplorer.Tests/G1ANT.Addon.IExplorer.Tests.csproj]: This change is not related to this task.

File: G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs (hunk 1 above)
Comment (lfronczyk) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: Check if the files already exist and don't extract them to speed-up Robot start.
If an exception is thrown display an error to the user (RobotMessageBox) and rethrow it.

File: G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs [hunk 4]
```diff
@@ -27,5 +26,37 @@ namespace G1ANT.Addon.Net
...
+        public override void LoadDlls()
+        {
+            UnpackDrivers();
+            base.LoadDlls();
+        }
+
+        private void UnpackDrivers()
+        {
+            var unpackfolder = AbstractSettingsContainer.Instance.UserDocsAddonFolder.FullName;
+            Dictionary<string, byte[]> exeList = new Dictionary<string, byte[]>()
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: `exeList` - the name is misleading: is it a driver (vide `UnpackDriver`) or an exe (vide `exeList`)? Is it a dictionary or a list? => rename to `drivers`.
Don't use full type name when it does not bring any useful information => var.

Did you verify the performance/memory penalty for referencing the resource? If there's any then you should rewrite the code to reference only the resources that need to be saved to user disk (see my next comment).

Comment (DanielWilczek) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: Its just copy pasted from other addon.
Than we need to change other addons also.

Comment (lfronczyk) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: > Its just copy pasted from other addon.

That's not a reason at all.

It would be great to update other addons, but it needs to be done in a separate task.

Comment (DanielWilczek) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: https://github.com/G1ANT-Robot/G1ANT.Addon/blob/develop/G1ANT.Addon.Selenium/G1ANT.Addon.Selenium/SeleniumAddon.cs

Comment (DanielWilczek) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: Made separate task for selenium addon: https://g1ant.manuscript.com/f/cases/7478/Addon-Selenium-normalize-resource-files-creation

**Comment by DanielWilczek** [comment]
All done

File: G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs [hunk 5]
```diff
@@ -27,5 +27,38 @@ namespace G1ANT.Addon.Net
...
+        {
+            var unpackfolder = AbstractSettingsContainer.Instance.UserDocsAddonFolder.FullName;
+            var driversDictionary = new Dictionary<string, byte[]>()
+            {
+                { "putty.exe", Resources.putty },
+                { "VNC.exe", Resources.VNC },
+                { "telnet.exe", Resources.telnet },
+            };
+            foreach (var exe in driversDictionary.Where(e => !File.Exists(Path.Combine(unpackfolder, e.Key))
+            || e.Value.Length !=new FileInfo(Path.Combine(unpackfolder, e.Key)).OpenRead().Length))
```
Comment (przegwar) [inline G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: CTRL+K CTRL+D

Comment (DanielWilczek) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: ????

File: G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs [hunk 6]
```diff
@@ -27,5 +27,38 @@ namespace G1ANT.Addon.Net
     [CommandGroup(Name = "vnc",  Tooltip = "A command connected with a VNC server.")]
     public class NetAddon : Language.Addon
     {
+        public override void Initialize()
```
Comment (pwitusik) [inline G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: unnecessary

Comment (DanielWilczek) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: done

File: G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs (hunk 1 above)
Comment (DanielWilczek) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: done

File: G1ANT.Addon.Net/G1ANT.Addon.Net/Commands/AS400OpenCommand.cs (hunk 2 above)
Comment (DanielWilczek) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/Commands/AS400OpenCommand.cs]: done

File: G1ANT.Addon.IExplorer/G1ANT.Addon.IExplorer.Tests/G1ANT.Addon.IExplorer.Tests.csproj (hunk 3 above)
Comment (DanielWilczek) [reply G1ANT.Addon.IExplorer/G1ANT.Addon.IExplorer.Tests/G1ANT.Addon.IExplorer.Tests.csproj]: reverted

File: G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs (hunk 5 above)
Comment (DanielWilczek) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: done

Comment (lfronczyk) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: `FileInfo` should already contain size of the file, you don't need to open it.
DRY - `Path.Combine(unpackfolder, e.Key)`; move the lambda (or some parts of it) into a method.
`driversDictionary` with `exe` elements => `driversDictionary` => `drivers`, `exe` => `driver`

Comment (DanielWilczek) [reply G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: done

File: G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs [hunk 7]
```diff
@@ -42,8 +42,7 @@ private void UnpackDrivers()
                 { "VNC.exe", Resources.VNC },
                 { "telnet.exe", Resources.telnet },
             };
-            foreach (var exe in driversDictionary.Where(e => !File.Exists(Path.Combine(unpackfolder, e.Key))
-            || e.Value.Length != new FileInfo(Path.Combine(unpackfolder, e.Key)).OpenRead().Length))
+            foreach (var exe in driversDictionary.Where(e => !DoesFileExist(unpackfolder, e.Key) || DoesFileSameLenght(e.Value.Length, unpackfolder, e.Key)))
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: DoesFileSameLenght => AreFilesOfTheSameLength

File: G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs [hunk 8]
```diff
@@ -55,5 +54,15 @@ private void UnpackDrivers()
...
         }
+
+        private bool DoesFileExist(string folder, string fileName)
+        {
+            return File.Exists(Path.Combine(folder, fileName));
+        }
+
+        private bool DoesFileSameLenght(int lenght, string folder, string fileName)
+        {
+            return lenght != new FileInfo(Path.Combine(folder, fileName)).Length;
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: length

**Comment by DanielWilczek** [comment]
@pwitusik @lfronczyk @przegwar Ping

File: G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs [hunk 9]
```diff
@@ -42,7 +42,7 @@ private void UnpackDrivers()
                 { "VNC.exe", Resources.VNC },
                 { "telnet.exe", Resources.telnet },
             };
-            foreach (var exe in driversDictionary.Where(e => !DoesFileExist(unpackfolder, e.Key) || DoesFileSameLenght(e.Value.Length, unpackfolder, e.Key)))
+            foreach (var exe in driversDictionary.Where(e => !DoesFileExist(unpackfolder, e.Key) || AreFilesOfTheSameLength(e.Value.Length, unpackfolder, e.Key)))
```
Comment (lfronczyk) [inline G1ANT.Addon.Net/G1ANT.Addon.Net/NetAddon.cs]: Same issues as in the other PR - probably missing negation before `AreFilesOfTheSameLength` + bad (unrelated) names for `exe` and `driversDictionary`.
